### PR TITLE
[cmake] Export sio library as target and cleanup cmake configuration

### PR DESCRIPTION
**Pull request by tmadlener**
BEGINRELEASENOTES
- Export sio library as `SIO::sio` target for easier downstream usage. (For the moment `SIO_INCLUDE_DIR` and `SIO_LIBRARIES` are also still defined in a call to `find_package(SIO)` to keep some backwards compatibility).
- Remove now obsolete cmake configuration by replacing it with more standard cmake approaches.
- Bump minimum required cmake version to 3.12

ENDRELEASENOTES

Up until now SIO did not export `sio` as a target, which makes it necessary to to something like
```cmake
add_library(some_lib)
target_link_libraries(some_lib ${SIO_LIBRARIES})
target_include_directories(some_lib ${SIO_INCLUDE_DIRS})
```

Now `sio` is exported in the `SIO` namespace, such that the single call to `target_link_libraries(some_lib SIO::sio)` should be enough. I also did some smaller cmake cleanups when defining this target.

Remi:
- [ ] Deal with ZLIB export ?
- [x] Export `sio-dump` in SIO CMake namespace?

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -56,5 +57,5 @@ ENDIF()
 # display some variables and write them to cache
 SIO_DISPLAY_STD_VARIABLES()
 
-# generate and install following configuration files
-SIO_GENERATE_PACKAGE_CONFIGURATION_FILES( SIOConfig.cmake SIOConfigVersion.cmake )
+### ADD CMAKE INFRASTRUCTURE FOR DOWNSTREAM #################################
+INCLUDE(cmake/SIOCreateConfig.cmake)
```
Comment (andresailer) [inline CMakeLists.txt:61]: Did you forget to include this SIOCreateConfig.cmake file?

Comment (tmadlener) [reply CMakeLists.txt:61]: I most certainly did. Sorry about that. Added it now.

**Comment by andresailer** [comment]
Thanks Thomas, I think the commit with the github action got lost with the force push?

File: source/CMakeLists.txt [hunk 2]
```diff
@@ -13,7 +13,6 @@ IF( NOT BIG_ENDIAN )
     ADD_DEFINITIONS( "-D__LITTLE_ENDIAN__" )
```
Comment (andresailer) [inline source/CMakeLists.txt]: Maybe use target_compile_definitions with PRIVATE/PUBLIC/INTERFACE depending on what is needed, then these are also forwarded, and the ADD_DEFINITIONS in SIOConfig could be removed.

**Comment by tmadlener** [comment]
> Thanks Thomas, I think the commit with the github action got lost with the force push?

Yes, I accidentally added that to the wrong branch. I have opened #16 with the changes, to keep the PRs more topical.

File: source/CMakeLists.txt (hunk 2 above)
Comment (tmadlener) [reply source/CMakeLists.txt]: @rete I suppose INTERFACE is the right choice here for the three definitions of `__LITTLE_ENDIAN__`, `SIO_LOGLVL` and `SIO_MACROS_WITH_EXCEPTIONS`? Or do I misunderstand the `INTERFACE` property once again?

Comment (rete) [reply source/CMakeLists.txt]: No you are right. This is where these variables should end up indeed. Can you add them too?

Comment (rete) [reply source/CMakeLists.txt]: Also check with a nice `make VERBOSE=1` that the flags are correctly passed during compilation

Comment (tmadlener) [reply source/CMakeLists.txt]: Done. Also tried briefly using the `SIO::sio` target in a downstream package and the definitions appear in the compiler invocations.

File: cmake/SIOConfig.cmake.in [hunk 3]
```diff
@@ -11,52 +12,34 @@
...
-
 
 # ---------- final checking ---------------------------------------------------
 INCLUDE( FindPackageHandleStandardArgs )
 # set SIO_FOUND to TRUE if all listed variables are TRUE and not empty
 # SIO_COMPONENT_VARIABLES will be set if FIND_PACKAGE is called with REQUIRED argument
-FIND_PACKAGE_HANDLE_STANDARD_ARGS( SIO DEFAULT_MSG SIO_ROOT SIO_INCLUDE_DIRS SIO_LIBRARIES ${SIO_COMPONENT_VARIABLES} ) #ZLIB_LIBRARIES )
+# Get the location from the target and ensure that it is present with the call below
+GET_PROPERTY(SIO_LIBRARY_TEST TARGET SIO::sio PROPERTY LOCATION)
+GET_PROPERTY(SIO_LIBRARIES TARGET SIO::sio PROPERTY NAME)
```
Comment (andresailer) [inline cmake/SIOConfig.cmake.in]: Why not more explicitly
```suggestion
SET(SIO_LIBRARIES SIO::sio)
```

**Comment by rete** [comment]
@tmadlener GH actions are ok. Can you rebase please?

**Comment by tmadlener** [comment]
For completeness: With this PR, `SIOConfig.cmake` and `SIOConfigVersion.cmake` will only be installed to `CMAKE_INSTALL_PREFIX/lib/cmake/SIO` but no longer to `CMAKE_INSTALL_PREFIX`, which was the case before. I have tried to build podio and LCIO against the version here, and setting `SIO_DIR` to the top level of the install directory (i.e. `CMAKE_INSTALL_PREFIX` of sio) was enough for picking up the right SIO library.

It would be possible to still install `SIOConfig.cmake`, `SIOConfigVersion.cmake`, as well as `SIOTargets.cmake` to the top level of `CMAKE_INSTALL_PREFIX`, to have exactly the same setup as before, but I am not sure it is actually necessary.

File: cmake/SIOCreateConfig.cmake [hunk 4]
```diff
@@ -0,0 +1,26 @@
+SET(CMAKE_INSTALL_CMAKEDIR "${CMAKE_INSTALL_LIBDIR}/cmake")
```
Comment (rete) [inline cmake/SIOCreateConfig.cmake:1]: Why having a separate file that you include only once in the top level CMakeLists.txt?

Comment (tmadlener) [reply cmake/SIOCreateConfig.cmake:1]: Mainly for a bit of separation between things that concern building and things that concern exporting things properly for usage in dependencies.

**Comment by rete** [comment]
It will not break anything for ilcsoft as only LCIO uses SIO and builds it internally if not found

File: source/CMakeLists.txt (hunk 2 above)
Comment (andresailer) [reply source/CMakeLists.txt]: I think PUBLIC would be the equivalent what was there before. With INTERFACE, SIO itself wouldn't be compiled with these defines, with PUBLIC SIO and things linking against it, with PRIVATE only SIO itself.

Comment (tmadlener) [reply source/CMakeLists.txt]: Yes, you are right it has to be PUBLIC. Changed.

**Review by rete — APPROVED** [review]

**Pull request merged**